### ISSUE TITLE
TIOModelTrainer class and TIOBatchDataSource protocol

### ISSUE THREAD

**Issue by phildow** [doc-ai/tensorio-ios#95]
Something like:

```objc
@protocol TIOBatchDataSource <NSObject>

- (NSUInteger)numberOfItems;
- (TIOBatchItem*)itemAtIndex:(NSUInteger)index;

@end
```

Where the batch data provider is responsible for vending data for training to a model trainer.

```objc
@interface TIOModelTrainer: NSObject

- (instancetype)initWithModel:(id<TIOTrainableModel>)model batch:(TIOBatch*)batch placeholders:(NSDictionary<NSString*, id<TIOData>>*)placeholders epochs:(NSUInteger)epochs batchSize:(NSUInteger)batchSize;

- (instancetype)initWithModel:(id<TIOTrainableModel>)model batchDataProvider:(id<TIOBatchDataProvider>)batchDataProvider placeholders:(NSDictionary<NSString*, id<TIOData>>*)placeholders epochs:(NSUInteger)epochs batchSize:(NSUInteger)batchSize;

- (NSDictionary<NSString*, id<TIOData>>*)train;

@end
```

Where the trainer takes either a fixed batch of data or a data provider along with a model and hyperparameters (placeholders, number of epochs, batch size) and iterates over the epochs, preparing batches from data the provider is vending, calling train on the underlying model for each pass.

In fact, if the trainer takes a fixed batch, it could instantiate itself as the data provider and then vend that fixed batched internally, so that the same API can be used. Or better it uses a utility object so as not to confuse consumers of the API.

Train then executes the training loop and when it is complete returns the final output value of the model, usually a loss value.

~

Tying into the federated learning and a FederatedTask API, the FederatedTask refers to a model and contains placeholders and hyperparameters. There is a question of how the data provider ties in (somewhere along the lines the app is registering itself for particular tasks and then vending data via the FederatedTask API), but otherwise it instantiates a model trainer given a task and model and executes it.
